controllers/cloudstackcluster_controller.go

```diff
@@ -69,52 +71,58 @@ func (r *CloudStackClusterReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	// Get CloudStack cluster.
 	csCluster := &infrav1.CloudStackCluster{}
-	if retErr = r.Client.Get(ctx, req.NamespacedName, csCluster); retErr != nil {
-		if client.IgnoreNotFound(retErr) == nil {
+	if err := r.Client.Get(ctx, req.NamespacedName, csCluster); err != nil {
+		if client.IgnoreNotFound(err) == nil {
 			log.Info("Cluster not found.")
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(retErr)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	// Get CAPI cluster.
-	cluster, retErr := util.GetOwnerCluster(ctx, r.Client, csCluster.ObjectMeta)
-	if retErr != nil {
-		return reconcile.Result{}, retErr
-	} else if cluster == nil {
+	capiCluster, err := util.GetOwnerCluster(ctx, r.Client, csCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	} else if capiCluster == nil {
 		log.Info("Waiting for CAPI Cluster controller to set owner reference on CloudStack cluster.")
 		return reconcile.Result{}, nil
 	}
 
 	// Check the cluster is not paused.
-	if annotations.IsPaused(cluster, csCluster) {
+	if annotations.IsPaused(capiCluster, csCluster) {
 		log.Info("Cluster is paused. Refusing to reconcile.")
 		return reconcile.Result{}, nil
 	}
 
 	// Setup patcher. This ensures modifications to the csCluster copy fetched above are patched into the origin.
-	if patchHelper, retErr := patch.NewHelper(csCluster, r.Client); retErr != nil {
-		return ctrl.Result{}, retErr
+	if patchHelper, err := patch.NewHelper(csCluster, r.Client); err != nil {
+		return ctrl.Result{}, err
 	} else {
 		defer func() {
-			if err := patchHelper.Patch(ctx, csCluster); retErr == nil && err != nil {
-				retErr = err
+			if err = patchHelper.Patch(ctx, csCluster); err != nil {
+				msg := "error patching CloudStackCluster %s/%s"
+				err = errors.Wrapf(err, msg, csCluster.Namespace, csCluster.Name)
+				retErr = multierror.Append(retErr, err)
 			}
 		}()
 	}
 
 	// Delete Cluster Resources if deletion timestamp present.
 	if !csCluster.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(log, csCluster)
+		return r.reconcileDelete(ctx, log, capiCluster, csCluster)
```

Comment:
I did some reading, and apparently context objects should always be the first argument in a function/method, and they should not be jammed into structs.

So, it's the first arg here.

Still, I'd like us to transition from passing so many args (not that bad in the cluster case, but the machine can get unwieldy)  to passing structs.

In this case it'd be something like `r.reconcileDelete(ctx, clstrCtrlrScope)`, but this conversion will be a bit of a chore to complete.

Reply:
I'm kinda on the other side of that argument.  I find listing the params individually in the function signature lets me see at a glimpse what's coming in, promoting grok-ability.  Yes, it's a bit more work to code/maintain, but I find it to be worth it.  

So I'm not a fan of *artificially* gathering params into structs.  (Of course, structs that define the objects of the app should be passed in as structs...)